File: sms-frontend/types/student_attendance.ts
// types/student-attendance.ts

export type AttendanceStatus =
    | "present"
    | "absent"
    | "late";



/* ========================================
   CREATE
======================================== */

export interface StudentAttendanceCreate {
    student_id: string;

    classroom_id?: string | null;

    attendance_date: string;

    status: AttendanceStatus;

    remarks?: string | null;
}



/* ========================================
   UPDATE
======================================== */

export interface StudentAttendanceUpdate {
    classroom_id?: string | null;

    attendance_date?: string;

    status?: AttendanceStatus;

    remarks?: string | null;
}



/* ========================================
   RESPONSE
======================================== */

export interface StudentAttendanceRead {
    id: string;

    student_id: string;

    classroom_id?: string | null;

    attendance_date: string;

    status: AttendanceStatus;

    remarks?: string | null;

    created_at: string;

    updated_at: string;
}



/* ========================================
   KPI
======================================== */
export interface AttendanceKPI {
    summary: {
        total_records: number;

        present: number;

        absent: number;

        late: number;

        attendance_rate: number;
    };

    monthly_attendance_chart: {
        month: string;
        attendance: number;
    }[];

    daily_attendance_chart: {
        day: string;
        attendance: number;
    }[];

    status_distribution: {
        present: number;

        absent: number;

        late: number;
    };

    top_absent_students: {
        student_name: string;

        total_absent: number;
    }[];

    classroom_performance: {
        classroom: string;

        attendance_rate: number;
    }[];
}

File: sms_back_end/database/multi_tenant_school_management/schemas/student_attendance.py
# schemas/student_attendance.py

from datetime import date, datetime
from typing import Optional
from uuid import UUID

from pydantic import BaseModel


# =========================
# BASE
# =========================

class StudentAttendanceBase(BaseModel):

    student_id: UUID

    classroom_id: Optional[UUID] = None

    attendance_date: date

    status: str = "present"

    remarks: Optional[str] = None


# =========================
# CREATE
# =========================

class StudentAttendanceCreate(
    StudentAttendanceBase
):
    pass


# =========================
# UPDATE
# =========================

class StudentAttendanceUpdate(BaseModel):

    classroom_id: Optional[UUID] = None

    attendance_date: Optional[date] = None

    status: Optional[str] = None

    remarks: Optional[str] = None


# =========================
# RESPONSE
# =========================

class StudentAttendanceResponse(
    StudentAttendanceBase
):

    id: UUID

    created_at: datetime

    updated_at: datetime

    model_config = {
        "from_attributes": True
    }